### PR TITLE
Fix Falcon tokenizer loading in pipeline

### PR DESCRIPTION
**Pull request by Rocketknight1**
Previously, pipelines checked `config.json` for info on the model tokenizer. However, often the information they want is not here, and is instead stored in `tokenizer_config.json`. This PR changes the pipeline code to more aggressively try `AutoTokenizer.from_pretrained()` even if `model.config` doesn't contain tokenizer information.

This PR is quite experimental for now, don't review/merge yet!

### CODE REVIEW

**Comment by Rocketknight1** [comment]
This turned out to be a more involved fix than I thought, so I'm going to put in a quick hack at @ArthurZucker's suggestion to fix it for Falcon, and come back to it later when the Keras 3 Situation is less urgent!

**Review by ArthurZucker — APPROVED** [review]
Thanks 😉

**Comment by Rocketknight1** [comment]
Some unrelated tests are red even after rebasing, will merge after that's resolved!

File: src/transformers/models/auto/tokenization_auto.py
Comment (ydshieh) [inline src/transformers/models/auto/tokenization_auto.py:164]: I am confused. The config 

https://huggingface.co/tiiuae/falcon-rw-1b/blob/main/tokenizer_config.json

says it use `GPT2Tokenizer`? Why we use `PreTrainedTokenizerFast` here?

I must miss something in the discussion.

Comment (ydshieh) [reply src/transformers/models/auto/tokenization_auto.py:164]: (I am surprised however put `PreTrainedTokenizerFast` here solved the issue we have, but if the original `Falcon` use `GPT2Tokenizer`, that's the value to be put here).

Comment (ArthurZucker) [reply src/transformers/models/auto/tokenization_auto.py:164]: yeah this was based on another model that used the pretrainedTokenizerFast. If GPT2TokenizerFast and GPT2Tokenizer work let's add them

Comment (Rocketknight1) [reply src/transformers/models/auto/tokenization_auto.py:164]: Hi @ydshieh @arthurzucker, the `falcon-rw` models are more like 'prototype' models and are quite different from the real models. The real models use `PreTrainedTokenizerFast`: https://huggingface.co/tiiuae/falcon-40b/blob/main/tokenizer_config.json

Comment (ydshieh) [reply src/transformers/models/auto/tokenization_auto.py:164]: But at the end, the real models still use a concrete tokenizer class. Do you know which class it is? Then you can put it into the auto mapping.

Comment (ydshieh) [reply src/transformers/models/auto/tokenization_auto.py:164]: (or is the type is actually `PreTrainedTokenizerFast` when we load it from the checkpoint ...? If so, this is going to blow my mind)

Comment (ydshieh) [reply src/transformers/models/auto/tokenization_auto.py:164]: Oh ............that's so strange. OK, then good for me this PR. 

But this is not something good at all and we need to see why this happens and fix it.

Can't really have such non-information (`PreTrainedTokenizerFast` in auto mapping nor in tokenizer_config.json): it brings too much confusion and doesn't provide explicit information.

Comment (ArthurZucker) [reply src/transformers/models/auto/tokenization_auto.py:164]: I think we can if we explain that there is pretty much no extra layer of python behind this class, but I am biased because it doesn't seem bad for me. It's actually good if we have a single class wrapping everyone 😈

Comment (ydshieh) [reply src/transformers/models/auto/tokenization_auto.py:164]: But if this holds, it means no one can answer simple question easily like which tokenizer a model uses, say `persimmon`.
We have auto tokenizer thtat abstract this, which is very good, but the information should exist somewhere IMO ..?

I am still surprised how a class like `PreTrainedTokenizerFast` could be used directly, but now it looks like `falcon` (the real one) doesn't use any tokenizer defined in `transformers` (i.e. the concrete class). 

Of course, if this is really the case (i.e. it doesn't belong to any tokenizer class), it makes sense. But it's a bit strange why it is the case.

Comment (ydshieh) [reply src/transformers/models/auto/tokenization_auto.py:164]: Anyway , feel free to merge as long as CI 💚 .

Comment (ydshieh) [reply src/transformers/models/auto/tokenization_auto.py:164]: BTW, @Rocketknight1 You are the author of the port no? So I guess you know the history 😄 . When you port the tokenizer for falcon at the beginning, you are not using any concrete tokenizer class as a reference?


Screenshot 2023-11-08 182816

Comment (Rocketknight1) [reply src/transformers/models/auto/tokenization_auto.py:164]: I just copied all of that from the falcon custom code! You can see it at the [last commit before I updated everything](https://huggingface.co/tiiuae/falcon-40b/tree/c47b371b31a68349c233104050ac76680b8485db).

Comment (ydshieh) [reply src/transformers/models/auto/tokenization_auto.py:164]: OK, I think they might do sth strange ... Thanks for the update.

**Pull request merged**
